File: main/src/main/java/org/anibyl/slounik/network/BatchArticlesLoader.java
package org.anibyl.slounik.network;

import android.content.Context;

import java.util.ArrayList;

/**
 * Batch loader of the dictionary site communicators.
 * <p/>
 * Created by Usievaład Kimajeŭ on 23.12.15.
 */
public class BatchArticlesLoader implements ArticlesLoader {
    private final DictionarySiteCommunicator[] communicators;

    public BatchArticlesLoader(DictionarySiteCommunicator... communicators) {
        this.communicators = communicators;
    }

    @Override
    public void loadArticles(String wordToSearch, Context context, final ArticlesCallback communicatorCallBack) {
        int activeCommunicators = 0;

        for (DictionarySiteCommunicator communicator : communicators) {
            if (communicator.enabled()) {
                activeCommunicators++;
                ArticlesCallback callback = new ArticlesCallback() {
                    @Override
                    public void invoke(ArticlesInfo info) {
                        ((BatchArticlesCallback) communicatorCallBack).invoke(this, info);
                    }
                };
                ((BatchArticlesCallback) communicatorCallBack).addCallback(callback);
                communicator.loadArticles(wordToSearch, context, callback);
            }
        }

        if (activeCommunicators == 0) {
            communicatorCallBack.invoke(new ArticlesInfo(ArticlesInfo.Status.FAILURE));
        }
    }

    public static abstract class BatchArticlesCallback extends ArticlesCallback {
        ArrayList<ArticlesCallback> callbacks = new ArrayList<>();

        void invoke(ArticlesCallback callback, ArticlesInfo info) {
            if (!callbacks.contains(callback)) {
                throw new RuntimeException("No such callback in batch callback.");
            }

            switch (info.getStatus()) {
                case SUCCESS:
                case FAILURE:
                    callbacks.remove(callback);

                    if (callbacks.size() != 0) {
                        info.setStatus(ArticlesInfo.Status.IN_PROCESS);
                    }

                case IN_PROCESS:
                    invoke(info);
                    break;
            }
        }

        public void addCallback(ArticlesCallback callback) {
            callbacks.add(callback);
        }
    }
}

File: main/src/main/java/org/anibyl/slounik/network/BatchArticlesLoader.kt
package org.anibyl.slounik.network

import android.content.Context
import java.util.ArrayList

/**
 * Batch loader of the dictionary site communicators.
 *
 * @author Usievaład Kimajeŭ
 * @created 23.12.2015
 */
class BatchArticlesLoader(private vararg val communicators: DictionarySiteCommunicator)
	: ArticlesLoader<BatchArticlesLoader.BatchArticlesCallback> {

	override fun loadArticles(wordToSearch: String, context: Context, callback: BatchArticlesCallback) {
		var activeCommunicators = 0

		for (communicator in communicators) {
			if (communicator.enabled()) {
				activeCommunicators++
				val articlesCallback = object : ArticlesCallback {
					override fun invoke(info: ArticlesInfo) {
						callback.invoke(this, info)
					}
				}
				callback.addCallback(articlesCallback)
				communicator.loadArticles(wordToSearch, context, articlesCallback)
			}
		}

		if (activeCommunicators == 0) {
			callback.invoke(ArticlesInfo(ArticlesInfo.Status.FAILURE))
		}
	}

	abstract class BatchArticlesCallback : ArticlesCallback {
		private var callbacks = ArrayList<ArticlesCallback>()

		internal operator fun invoke(callback: ArticlesCallback, info: ArticlesInfo) {
			if (!callbacks.contains(callback)) {
				// Should not happen, however, I had couple occurrences in production.
				return
			}

			when (info.status) {
				ArticlesInfo.Status.SUCCESS, ArticlesInfo.Status.FAILURE -> {
					callbacks.remove(callback)

					if (callbacks.size != 0) {
						info.status = ArticlesInfo.Status.IN_PROCESS
					}
					invoke(info)
				}

				ArticlesInfo.Status.IN_PROCESS -> invoke(info)
			}
		}

		fun addCallback(callback: ArticlesCallback) {
			callbacks.add(callback)
		}
	}
}
